### ISSUE TITLE
Macros support

### ISSUE THREAD

**Issue by fuson** [KronicDeth/intellij-elixir#109]
imho It will be the most useful feature, cause top solutions (like Phoenix) massively use this. DSL is one of the core ideas of Elixir.

**Comment by KronicDeth** [comment]
I had always planned to support macros by implementing the full parser and AST (quoted) output instead of writing a partial parser that just pretends that `defmodule` and `def` are keywords as some other editor plugins do.

I can actually spit out the quoted AST for the current rules the parser supports (see the `quote` methods in [ElixirPsiimplUtil](https://github.com/KronicDeth/intellij-elixir/blob/35a78fd605e0d41591501f5e538c146c74d35988/src/org/elixir_lang/psi/impl/ElixirPsiImplUtil.java) and use this in the [parser tests](https://github.com/KronicDeth/intellij-elixir/blob/35a78fd605e0d41591501f5e538c146c74d35988/tests/org/elixir_lang/parser_definition/ParsingTestCase.java) to compare my parsers quoted form to the native Elixir quoted form using JInterface and [intellij_elixir](https://github.com/intellij_elixir).

I could probably expose the quoted form using the virtual PSIElement support that IntelliJ normally uses to show variable values (in green to the right of the actual code) during debugging like this:
![screen shot 2015-04-08 at 10 39 04 pm](https://cloud.githubusercontent.com/assets/298259/7060068/2ed324c0-de40-11e4-9f90-5e0afb22935d.png)

Is this part of what you're thinking of when you say "Macros support"?  "Macros support" is a very broad topic and I would classify it as an Epic in Agile terms.  I'd like you to enumerate smaller features, like showing the quoted form as I've outlined that can be separated into their own issues so they can be delivered and closed quicker than this issue which covers all work related to macros.

Other features that will already apply to macros:
* [References](https://github.com/KronicDeth/intellij-elixir/issues/93) - general feature needed to support Go To Symbol and Find Usages
* [Go To Symbol](https://github.com/KronicDeth/intellij-elixir/issues/96) - Cmd+Click a macro name and go to it's `defmacro`
* [Find Usages](https://github.com/KronicDeth/intellij-elixir/issues/94) - Right-click a defmacro or a macro name and find all the places it is used.
* [Structure View](https://github.com/KronicDeth/intellij-elixir/issues/97) - The Structure pane will show macros as distinct from modules and functions.

**Comment by rambocoder** [comment]
Ideally it would be awesome if macros had a different color when looking at the source code of the ex file where they are used, even though that might be kind of redundant because most of elixir code is macros :)

**Comment by KronicDeth** [comment]
From my understanding of the [Custom Language Guide](https://confluence.jetbrains.com/display/IntelliJIDEA/Syntax+Highlighter+and+Color+Settings+Page), the [`lang.syntaxHighlighterFactory` extensionPoint](https://github.com/JetBrains/intellij-community/blob/d3a415a9c03c8a93aaa6c69f1abdea4f7089f5be/platform/platform-resources/src/META-INF/PlatformExtensionPoints.xml#L86-L89) I'm using for the [current syntax highlighting](https://github.com/KronicDeth/intellij-elixir/blob/f7fe71a8df35211416a093ad1f408228a2220765/META-INF/plugin.xml#L169) only works on [lexer](https://github.com/KronicDeth/intellij-elixir/blob/f7fe71a8df35211416a093ad1f408228a2220765/src/org/elixir_lang/Elixir.flex) tokens, which are below the level where the plugin can distinguish between functions and macros, but looking through the [available extension points](https://github.com/JetBrains/intellij-community/blob/d3a415a9c03c8a93aaa6c69f1abdea4f7089f5be/resources/src/idea/RichPlatformPlugin.xml#L23-L89) for IntelliJ, there is a [`highlightVisitor` extensionPoint](https://github.com/JetBrains/intellij-community/blob/d3a415a9c03c8a93aaa6c69f1abdea4f7089f5be/platform/platform-resources/src/META-INF/LangExtensionPoints.xml#L30-L32) that is used in IntelliJ to [apply other, higher level highlighting like annotations](https://github.com/JetBrains/intellij-community/blob/d3a415a9c03c8a93aaa6c69f1abdea4f7089f5be/platform/analysis-impl/src/com/intellij/codeInsight/daemon/impl/DefaultHighlightVisitor.java) that I may be able to use to highlight macros after the syntax highlighting pass.  If you have any Java background, @rambocoder, I could use the help find examples from JetBrains/intellij-community, JetBrains/intellij-plugins or other open source plugins for Jetbrains to see if there is some plugin that does a similar second-pass highlighting.  In intellij-erlang I assume macro highlighting can be performed at the lexer level because erlang macros are prefixed with `?`.

**Comment by raarts** [comment]
This is probably also the reason why I can't step through tests with the debugger?

**Comment by KronicDeth** [comment]
@raarts no.  It has nothing to do with macros.  It's because test files don't generate `.beam` files, which are needed for the Erlang debugging libraries to work.
